### PR TITLE
apps: charts to ghcr

### LINKED ISSUE #1805: [5] Generate helm chart repository to simplify sharing of charts
**Is your feature request related to a problem? Please describe.**
Some charts here in apps is reused in other repos to integrate with apps, or to use the same type of policies as we do here.
Examples such as networkpolicies, podsecuritypolicies, and support and utilities for running opensearch. And we can definitely improve this situation to make it simpler for this sharing to happen.

**Describe the solution you'd like**
Either that the charts are packaged and pushed to an OCI registry, or our own Harbor, or that we generate an index and do a GitHub pages site to act as a repository.

**Additional context**
This would also imply that repos charts from here add automation to simply chart updates there.

**Definition of done:**
Our charts in apps are accessible via a repository.

### PR DESCRIPTION
**Pull request by robinelastisys**
> [!warning]
> **This is public repository, ensure not to disclose:**
> - [x] personal data beyond what is necessary for interacting with this pull request
> - [x] business confidential information, such as customer names

### What kind of PR is this?

**Required**: Mark one of the following that is applicable:

- [x] kind/feature       
- [ ] kind/improvement   
- [ ] kind/deprecation   
- [ ] kind/documentation 
- [ ] kind/clean-up      
- [ ] kind/bug           
- [ ] kind/other         

*Optional*: Mark one or more of the following that are applicable:

> [!important]
> Breaking changes should be marked `kind/admin-change` or `kind/dev-change` depending on type
> Critical security fixes should be marked with `kind/security`

- [ ] kind/admin-change  
- [ ] kind/dev-change    
- [ ] kind/security      
- [ ] [kind/adr]()       

### What does this PR do / why do we need this PR?

...

- Fixes #1805

#### Checklist

- [x] Proper commit message prefix on all commits
  
- Change checks:
  - [x] The change is transparent
  - [ ] The change is disruptive
  - [ ] The change requires no migration steps
  - [ ] The change requires migration steps
- Metrics checks:
  - [x] The metrics are still exposed and present in Grafana after the change
  - [ ] The metrics names didn't change (Grafana dashboards and Prometheus alerts are not affected)
  - [ ] The metrics names did change (Grafana dashboards and Prometheus alerts were fixed)
- Logs checks:
  - [x] The logs do not show any errors after the change
- Network Policy checks:
  - [x] Any changed pod is covered by Network Policies
  - [ ] The change does not cause any dropped packages in the `NetworkPolicy Dashboard`
- Pod Security Policy checks:
  - [x] Any changed pod is covered by Pod Security Admission
  - [ ] Any changed pod is covered by Gatekeeper Pod Security Policies
  - [ ] The change does not cause any pods to be blocked by Pod Security Admission or Policies
- Falco checks:
  - [x] The change does not cause any alerts to be generated by Falco
- Audit checks:
  - [x] The change does not cause any unnecessary Kubernetes audit events
  - [ ] The change requires changes to Kubernetes audit policy
- Bug checks:
  - [x] The bug fix is covered by regression tests

### CODE REVIEW

File: .github/workflows/push-to-ghcr.yml [hunk 1]
```diff
@@ -0,0 +1,35 @@
+name: Push package to GHCR
+
+env:
+  DOTNET_VERSION: '6.0.x'
```
Comment (Xartos) [inline .github/workflows/push-to-ghcr.yml:4]: Why is this needed?

File: .github/workflows/push-to-ghcr.yml [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+          - name: Import GPG key
+            uses: crazy-max/ghaction-import-gpg@v6
+            with:
+              gpg_private_key: ${{ secrets.HELM_SIGN_PGP_PRIVATE_KEY }}
+              passphrase: ''
+
+          - name: Export PGP Private Key
+            run: |
+              gpg --export-secret-keys 'Helm Package Signing' > ~/.gnupg/private.gpg
```
Comment (Xartos) [inline .github/workflows/push-to-ghcr.yml:24]: Do you need to export the key if it's already in the keyring? Like if you remove this and the `--keyring ~/.gnupg/private.gpg` flag from the helm command.

File: .github/workflows/push-to-ghcr.yml [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+          - name: Helm Registry Login
+            run: helm registry login ghcr.io -u ${{ github.repository_owner }} --password ${{ secrets.GITHUB_TOKEN }}
+
+          - name: Package, Sign And Push To Registry
```
Comment (Xartos) [inline .github/workflows/push-to-ghcr.yml:29]: You could maybe use a matrix here instead then you could have a yaml list and skip the for loop in the bash part to make it a bit easier to read https://code.dblock.org/2021/09/03/generating-task-matrix-by-looping-over-repo-files-with-github-actions.html

File: .github/workflows/push-to-ghcr.yml (hunk 2 above)
Comment (robinelastisys) [reply .github/workflows/push-to-ghcr.yml:24]: Yeah, sadly need to ensure it is the right key format to make it work! This export make sure it is.

File: .github/workflows/push-to-ghcr.yml (hunk 3 above)
Comment (robinelastisys) [reply .github/workflows/push-to-ghcr.yml:29]: Will have a look! :+1:

**Review by aarnq — COMMENTED** [review]
I even get _chart signing_? :smile: 

We should document this, how to consume the charts, and how to verify the charts.

Additionally we should enforce a versioning policy of the charts.

But maybe I should work on that policy in a separate PR, as we also need to do that in some clever way for images as well.

**Comment by robinelastisys** [comment]
> I even get _chart signing_? 😄
> 
> We should document this, how to consume the charts, and how to verify the charts.
> 
> Additionally we should enforce a versioning policy of the charts.
> 
> But maybe I should work on that policy in a separate PR, as we also need to do that in some clever way for images as well.

Yep! Signing it is!

File: .github/workflows/push-to-ghcr.yml [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+          gpg --export-secret-keys 'helm sign' > ~/.gnupg/private.gpg
+
+      - name: Helm Registry Login
+        run: helm registry login ghcr.io -u ${{ github.repository_owner }} --password ${{ secrets.GITHUB_TOKEN }}
+
+      - name: Package, Sign And Push To Registry
+        run: |
+          folder="${{ matrix.folder }}"
+          if [ -d "$folder" ]; then
+            helm package --sign --key "helm sign" --keyring ~/.gnupg/private.gpg "$folder" -d "$folder" && helm push "$folder"/*.tgz oci://ghcr.io/${{ github.repository_owner }} || true
```
[the remark this answers is not in the document]
Comment (robinelastisys) [reply .github/workflows/push-to-ghcr.yml]: Apparently not anymore! :man_shrugging:  Removed!

File: .github/workflows/push-to-ghcr.yml [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+          - './helmfile.d/charts/openstack-monitoring'
+          - './helmfile.d/charts/gatekeeper/templates'
+          - './helmfile.d/charts/gatekeeper/mutations'
+          - './helmfile.d/charts/gatekeeper/metrics'
+          - './helmfile.d/charts/gatekeeper/constraints'
+          - './helmfile.d/charts/gatekeeper/podsecuritypolicies'
+          - './helmfile.d/charts/networkpolicy/generator'
+          - './helmfile.d/charts/networkpolicy/common'
+          - './helmfile.d/charts/networkpolicy/workload-cluster'
+          - './helmfile.d/charts/networkpolicy/service-cluster'
```
[the remark this answers is not in the document]
Comment (aarnq) [reply .github/workflows/push-to-ghcr.yml]: I think you were too quick here, we need to keep the OpenSearch charts in this list.

File: helmfile.d/charts/README.md [hunk 6]
```diff
@@ -0,0 +1,24 @@
+# GHCR
+
+As part of our effort to make our helmcharts available for public and internal consumption we have decided to publicise some of them in our "Github Container Registry".
```
Comment (aarnq) [inline helmfile.d/charts/README.md:3]:
```suggestion
As part of our effort to make our Helm charts available for public and internal consumption we have decided to publish some of them on Github Container Registry.
```

File: .github/workflows/push-to-ghcr.yml [hunk 7]
```diff
@@ -0,0 +1,45 @@
+name: Push package to GHCR
+
+on:
+  push:
```
Comment (aarnq) [inline .github/workflows/push-to-ghcr.yml:4]: Should we limit this to push on main?

File: .github/workflows/push-to-ghcr.yml [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+
+jobs:
+  push-store-image:
```
Comment (aarnq) [inline .github/workflows/push-to-ghcr.yml:7]:
```suggestion
  publish-chart:
```

File: .github/workflows/push-to-ghcr.yml (hunk 7 above)
Comment (robinelastisys) [reply .github/workflows/push-to-ghcr.yml:4]: I had this setup just to see it work.. Now when we know it works it should be enough with the main branch right?

File: helmfile.d/charts/helmverify.gpg
Comment (aarnq) [inline helmfile.d/charts/helmverify.gpg:1]: If we are going to provide this as a token that "this is issued from us" then we must have it so the user information of the key is "Elastisys Compliant Kubernetes - Chart Signing Key" and (I assume) our support mail unless @Xartos has a different opinion.

Comment (robinelastisys) [reply helmfile.d/charts/helmverify.gpg:1]: Done! See if u like it! :+1:

**Review by aarnq — APPROVED** [review]

**Review by Xartos — APPROVED** [review]
